aws-java-sdk-core/src/main/java/com/amazonaws/auth/profile/internal/ProfilesConfigFileLoader.java

```diff
@@ -82,43 +85,99 @@
-                                    new BasicSessionCredentials(accessKey, secretKey, sessionToken)));
-            }
+            return new Profile(profileName,
+                    new BasicSessionCredentials(accessKey, secretKey, sessionToken));
         }
+    }
 
-        return profilesByName;
+    private static Profile fromAssumeRole(String profileName, Map<String, String> properties, Map<String, Map<String, String>> allProfileProperties, ProfileCredentialsService profileCredentialsService) {
+        String roleArn = properties.get(Profile.ROLE_ARN);
+        String sourceProfileName = properties.get(Profile.SOURCE_PROFILE);
+        String roleSessionName = properties.get(Profile.ROLE_SESSION_NAME);
+        String externalId = properties.get(Profile.EXTERNAL_ID);
+
+        if (sourceProfileName == null) {
+            throw new AmazonClientException(
+                    String.format("Unable to load credentials into profile [%s]: Source profile name is not specified",
+                            profileName));
+        }
+
+        Map<String, String> sourceProfileProperties = allProfileProperties.get(sourceProfileName);
+
+        if (sourceProfileProperties == null) {
+            throw new AmazonClientException(
+                    String.format("Unable to load source profile [%s]: Source profile not found [%s]",
+                            profileName, sourceProfileName));
+        }
+
+        String sourceAccessKey = sourceProfileProperties.get(Profile.AWS_ACCESS_KEY_ID);
+        String sourceSecretKey = sourceProfileProperties.get(Profile.AWS_SECRET_ACCESS_KEY);
+        String sourceSessionToken = sourceProfileProperties.get(Profile.AWS_SESSION_TOKEN);
+        AWSCredentials sourceCredentials;
+
+        if (sourceAccessKey == null) {
+            throw new AmazonClientException(
+                    String.format("Unable to load credentials into profile [%s]: AWS Access Key ID is not specified in source profile [%s].",
+                            profileName, sourceProfileName));
+        }
+
+        if (sourceSecretKey == null) {
+            throw new AmazonClientException(
+                    String.format("Unable to load credentials into profile [%s]: AWS Secret Access Key is not specified in source profile [%s].",
+                            profileName, sourceProfileName));
+        }
+
+        if (sourceSessionToken == null) {
+            sourceCredentials = new BasicAWSCredentials(sourceAccessKey, sourceSecretKey);
+        } else {
+            sourceCredentials = new BasicSessionCredentials(sourceAccessKey, sourceSecretKey, sourceSessionToken);
+        }
+
+        if (roleSessionName == null) {
+            roleSessionName = "aws-sdk-java-" + System.currentTimeMillis();
+        }
+
+        RoleInfo roleInfo = new RoleInfo(roleArn, roleSessionName, externalId);
+        return new Profile(profileName, sourceProfileName, profileCredentialsService.assumeRole(sourceCredentials, roleInfo), roleInfo);
```

Comment:
If I'm reading the code correctly it seems we will try and assume the role for every profile we find regardless of whether it's used or not. It's odd that each ProfileCredentialsProvider will parse and load all profiles each time. I think I'd prefer we have a single instance of profile metadata that doesn't try to do anything with the data (i.e. no assuming roles) and when a ProfileCredentialsProvider is created it would reach out to this static registry of profile data (lazily loaded), get the information in the profile and from there do something with it (i.e. create credentials object, assume role, etc). That way we only throw an exception when a customer is using a assume role profile without having STS rather then just having it in their profile file but not actually reference it in the Java SDK (which I imagine would be pretty common since it works for the CLI and they could just be using different profiles for different tools)